packages/ts-moose-lib/src/streaming-functions/runner.ts

```diff
@@ -623,69 +480,41 @@ const sendMessages = async (
   producer: Producer,
   messages: KafkaMessageWithLineage[],
 ): Promise<void> => {
-  try {
-    let chunk: KafkaMessageWithLineage[] = [];
-    let chunkSize = 0;
-
-    const maxMessageSize = targetTopic.max_message_bytes || 1024 * 1024;
+  if (messages.length === 0) return;
 
-    for (const message of messages) {
-      const messageSize =
-        Buffer.byteLength(message.value, "utf8") +
-        KAFKAJS_BYTE_MESSAGE_OVERHEAD;
-
-      if (chunkSize + messageSize > maxMessageSize) {
-        logger.log(
-          `Sending ${chunkSize} bytes of a transformed record batch to ${targetTopic.name}`,
-        );
-        // Send the current chunk before adding the new message
-        await sendChunkWithRetry(
-          logger,
-          targetTopic,
-          producer,
-          chunk,
-          maxMessageSize,
-        );
-        logger.log(
-          `Sent ${chunk.length} transformed records to ${targetTopic.name}`,
-        );
+  try {
+    // Library handles batching and retries internally
+    await producer.send({
+      topic: targetTopic.name,
+      messages: messages,
+    });
 
-        // Start a new chunk
-        chunk = [message];
-        chunkSize = messageSize;
-      } else {
-        // Add the new message to the current chunk
-        chunk.push(message);
-        metrics.bytes += Buffer.byteLength(message.value, "utf8");
-        chunkSize += messageSize;
-      }
+    // Track metrics only after successful send to target topic
+    // Messages routed to DLQ should NOT be counted as successful sends
+    for (const msg of messages) {
+      metrics.bytes += Buffer.byteLength(msg.value, "utf8");
     }
+    metrics.count_out += messages.length;
 
-    metrics.count_out += chunk.length;
-
-    // Send the last chunk
-    if (chunk.length > 0) {
-      logger.log(
-        `Sending ${chunkSize} bytes of a transformed record batch to ${targetTopic.name}`,
-      );
-      await sendChunkWithRetry(
-        logger,
-        targetTopic,
-        producer,
-        chunk,
-        maxMessageSize,
-      );
-      logger.log(
-        `Sent final ${chunk.length} transformed data to ${targetTopic.name}`,
-      );
-    }
+    logger.log(`Sent ${messages.length} messages to ${targetTopic.name}`);
   } catch (e) {
+    // Library already retried - this is a permanent failure
     logger.error(`Failed to send transformed data`);
     if (e instanceof Error) {
       logError(logger, e);
     }
-    // This is needed for retries
-    throw e;
+
+    // Handle DLQ for failed messages
+    // Only throw if not all messages were successfully routed to DLQ
+    const allHandledByDLQ = await handleDLQForFailedMessages(
+      logger,
+      producer,
+      messages,
+      e,
+    );
+    if (!allHandledByDLQ) {
+      throw e;
```

Comment:
i see we used to throw for retries but since that is handled by the kafka library do we still need to?

Reply:
Yes - for at least once delivery